### PR TITLE
ML rules fixes, new rule for msgpack-numpy

### PR DESCRIPTION
**Pull request by dhalf**
This PR contains:
- Some changes/fixes to ML-related Python rules and/or their associated test file: lxml-in-pandas, numpy-in-pytorch-modules and pickles-in-pytorch
- A new rule to spot the usage of msgpack-numpy, a Python library that prepares numpy objects to be serialized with msgpack but that is backed by pickle. The rule excludes normal uses of msgpack (without the numpy translation layer).

### CODE REVIEW

**Comment by GrosQuildu** [comment]
Thanks! Will review soon. Seems like you have to sign the CLA?

**Comment by dhalf** [comment]
> Thanks! Will review soon. Seems like you have to sign the CLA?

Hi! I did sign twice but this doesn't seem to be reflected here... Any idea?

**Comment by dhalf** [comment]
I added two new rules to detect the use of Keras and Tensorflow loading functions. We reject all calls except those with a string literal (or constant) as file path. I also added a warning rule to deprecate the usage of pickle and hdf5 for Keras. The preferred format is now .keras which comes with checks that mitigate the risk of arbitrary code execution.

**Comment by GrosQuildu** [comment]
Overall looks good. Need @suhacker1 to review these (or somebody from ML). Then I can review from the semgrep/cq perspective.

**Review by suhacker1 — DISMISSED** [review]
From the ML-perspective, everything looks good to me!

**Review by GrosQuildu — DISMISSED** [review]

File: python/pandas-eval.yaml [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+    patterns:
+      - pattern-inside: |
+          import pandas
+          ...
+      - pattern-either:
+          - patterns:
+              - pattern: pandas.DataFrame.$FN(...)
+              - pattern-not: pandas.DataFrame.$FN("...", ...)
+              - pattern-not: pandas.DataFrame.$FN(f"", ...)
```
Comment (mschwager) [inline python/pandas-eval.yaml:31]: Is `f""` intended? We're trying to remove only empty f-strings?

File: python/numpy-in-pytorch-modules.yaml [hunk 2]
```diff
@@ -15,7 +15,9 @@ rules:
         - https://tanelp.github.io/posts/a-bug-that-plagues-thousands-of-open-source-ml-projects
 
     patterns:
-      - pattern: $RESULT = numpy.$FUNCTION(...)
+      - pattern-either:
+          - pattern: numpy.$FN(...)
+          - pattern: numpy.$MOD.$FN(...)
```
Comment (mschwager) [inline python/numpy-in-pytorch-modules.yaml:20]: You can also use something like `numpy. ... .$FN(...)` if you'd like to match arbitrary depth.

File: python/pickles-in-keras-deprecation.yaml [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+      - pattern-not:
+          patterns:
+            - pattern-either:
+                - pattern: keras.models.load_model($FILE)
+                - pattern: tensorflow.keras.models.load_model($FILE)
+                - pattern: keras.saving.load_model($FILE)
+                - pattern: tensorflow.keras.saving.load_model($FILE)
+            - metavariable-comparison:
+                metavariable: $FILE
+                comparison: re.match(".*\.keras", $FILE)
```
Comment (mschwager) [inline python/pickles-in-keras-deprecation.yaml:36]: I think this is great as-is. I tend to write rules of this form like this, which may be slightly more ergonomic:

```yaml
                - pattern: tensorflow.keras.saving.load_model("$FILE")
            - metavariable-regex:
                metavariable: $FILE
                regex: ".*\.keras"
```

**Review by mschwager — COMMENTED** [review]
LGTM! I left a few minor comments 👍

File: python/pandas-eval.yaml (hunk 1 above)
Comment (GrosQuildu) [reply python/pandas-eval.yaml:31]: It is intended. We want to remove all static strings, including empty strings. Semgrep seems to have a bug - according to semgrep:
* `f" expr "` - is static string
* `f" {expr} "` - is dynamic string
* `f""` - is dynamic string

Will add a testcase for that and make issue in the upstream r2c

File: python/numpy-in-pytorch-modules.yaml (hunk 2 above)
Comment (GrosQuildu) [reply python/numpy-in-pytorch-modules.yaml:20]: Good point, fixed

File: python/pickles-in-keras-deprecation.yaml (hunk 3 above)
Comment (GrosQuildu) [reply python/pickles-in-keras-deprecation.yaml:36]: fixed

File: python/pandas-eval.yaml (hunk 1 above)
Comment (GrosQuildu) [reply python/pandas-eval.yaml:31]: https://github.com/semgrep/semgrep/issues/10047

Comment (GrosQuildu) [reply python/pandas-eval.yaml:31]: Issue fixed in upstream, we can fix that now.

**Pull request merged**